File: src/bot/GraphBot.java
package bot;

import java.util.Deque;
import java.util.HashSet;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;

import gameLogic.Brain;
import gameLogic.Direction;
import gameLogic.GameState;
import gameLogic.Position;
import gameLogic.Snake;

public class GraphBot implements Brain
{
	private final static int MAX_DEPTH = 256;
	private Snake self;
	private GameState state;
	private Graph graph;
	private BoardAnalyzer analyzer;
	private long startTime;
	
	public GraphBot()
	{
		System.out.println();
	}

	@Override
	public Direction getNextMove(Snake yourSnake, GameState gameState)
	{
		this.startTime = System.currentTimeMillis();
		this.self = yourSnake;
		this.state = gameState;
		this.analyzer = new BoardAnalyzer(state, self);
		this.graph = new Graph(analyzer);
		
		final Position from = self.getHeadPosition();
		Deque<Direction> path = null;
		while(path == null)
		{	
			final Position to = analyzer.getNextTarget();
			path = graph.getBfsPath(from, to);
		}
		
		Direction firstDirection = path.peek();
		if(isOpositeDirections(self.getCurrentDirection(), firstDirection))
			firstDirection = firstDirection.turnLeft();
		
		final Direction direction = getBestDirection(firstDirection, from);
		return direction;
	}

	private Direction getBestDirection(Direction firstDirection, final Position from)
	{
		SortedMap<Integer, Direction> scores = new TreeMap<Integer, Direction>();
		Set<Position> visited = new HashSet<Position>(128);
		visited.add(from);
		
		
		Position next = firstDirection.calculateNextPosition(from);
		int score = getScore(visited, next, 0);
		scores.put(score, firstDirection);
		
		for(Direction d : Direction.values())
		{
			if(d.equals(firstDirection))
				continue;
			next = d.calculateNextPosition(from);
			score = getScore(visited, next, 0);
			scores.put(score, d);
		}
		
		final int highScore = scores.lastKey();
		final Direction bestDirection = scores.get(highScore);
		System.out.println(bestDirection + "("+highScore+")");
		
		return scores.get(highScore);
	}

	private int getScore(Set<Position> visited, Position current, int score)
	{
		if(visited.contains(current))
			return score;
		
		if(reachedComputationCapacity(visited.size()))
			return score;
		
		if(analyzer.isLethal(current))
			return score;		
		
		visited.add(current);
		
		score += analyzer.getScore(current);
		
		int highScore = Integer.MIN_VALUE;
		
		for(Direction d : Direction.values())
		{
			final Position next = d.calculateNextPosition(current);
			final int scoreInDirection = getScore(visited, next, score);
			if(scoreInDirection > highScore)
				highScore = scoreInDirection;
		}
		visited.remove(current);
		
		return highScore;
	}
	
	private boolean reachedComputationCapacity(final int pathSize)
	{
		if(thinkingTimeLeft() < 10)
			return true;

		if(pathSize >= MAX_DEPTH)
			return true;
		
		return false;
	}
	
	private long thinkingTimeElapsed()
	{
		return System.currentTimeMillis() - startTime;
	}

	private long thinkingTimeLeft()
	{
		final long elpasedTime = thinkingTimeElapsed();
		return state.getMetadata().getMaximumThinkingTime() - elpasedTime;
	}

	private boolean isOpositeDirections(Direction currentDireciton, Direction dir)
	{
		dir = dir.turnLeft().turnLeft();
		return currentDireciton.equals(dir);
	}
}

File: src/main/kotlin/bot/GraphBot.kt
package bot

import gameLogic.*
import java.util.*
import kotlin.collections.HashSet


class GraphBot: Brain
{
	private var memory: Matrix<Double> = Matrix(1, 1, 0.0)
	private val maxDepth = 256
	init
	{
		System.out.println("\n\n*** BAM!! ***\n")
		System.out.println("*** Version ${this::class.java.hashCode()} ***\n")
	}

	override fun getNextMove(self: Snake, gameState: GameState): Direction
	{
		try
		{
			if(memory.size == 1)
				memory = Matrix<Double>(gameState.board)

			println("Before forgetting: $memory")
			forget(memory)
			memory[self.headPosition] = -1.0

			val time = TimeTracker(gameState.metadata)
			val state: BoardState = BoardState(gameState, self)
			val graph = Graph(state)

			val fruits = rankFruits(state, self.headPosition)
			val firstDirectionToEvaluate = initialDirection(graph, fruits, state, self)
			val from = self.headPosition
			val direction = bestDirection(firstDirectionToEvaluate, self.currentDirection, from, state, time, memory)
			return direction
		}
		catch(e: Exception)
		{
			e.printStackTrace()
		}

		return self.currentDirection.turnLeft()
	}

	private fun forget(memory: Matrix<Double>)
	{
		memory.indices.asSequence()
				.filter { memory[it] != 0.0 }
				.forEach { memory[it] = reduce(memory[it]) }
	}

	private fun reduce(i: Double): Double = i/10.0

	private fun bestDirection(initialDirection: Direction, currentDirection: Direction, current: Position, state: BoardState, time: TimeTracker, memory: Matrix<Double>): Direction
	{
		val directions: Queue<Direction> = listOfDirections(initialDirection, currentDirection)
		val scores = TreeMap<Int, Direction>()
		val visited: MutableSet<Position> = HashSet(maxDepth /4)

		while(directions.isNotEmpty())
		{
			val nextDirection: Direction = directions.poll()
			val nextPosition: Position = current going nextDirection
			val score = getScore(visited, nextPosition, state, time, memory)
			scores.put(score, nextDirection)
		}

		val highScore: Int = scores.lastKey()
		val bestDirection: Direction = scores[highScore] ?: initialDirection
		println("${this::class.java.canonicalName}: $bestDirection ($highScore)")

		return bestDirection
	}

	/**
	 * Create a list of all Direction values, where the initialDirection is first in the list.
	 */
	private fun listOfDirections(initialDirection: Direction, currentDirection: Direction): Queue<Direction>
	{
		val q: Queue<Direction> = LinkedList()
		val complement = Direction.values()
				.filter { it != initialDirection }
				.filter { it != currentDirection }

		q.add(initialDirection)
		q.addAll(complement)

		return q
	}

	private fun getScore(visited: MutableSet<Position>, current: Position, state: BoardState, time: TimeTracker, memory: Matrix<Double>, score: Int = 0): Int
	{
		if(current in visited)
			return score

		if(reachedComputationCapacity(time, visited.size))
			return score

		if(state.isLethal(current))
			return score - 10

		visited.add(current)

		val scoreForThis: Byte = state.score(current)
		val accumulatedScore: Int = score + scoreForThis
		val highScore: Int = directions(current, memory)
				.asSequence()
				.map { current going it }
				.map { getScore(visited, it, state, time, memory, accumulatedScore) }
				.max() ?: accumulatedScore

		memory[current] = updateMemory(current, memory, highScore)
		visited.remove(current)

		return highScore
	}

	private fun updateMemory(current: Position, memory: Matrix<Double>, accumulatedScore: Int): Double
	{
		val inMemory = memory[current]

		if(Math.abs(inMemory) > Math.abs(accumulatedScore))
			return inMemory
		return accumulatedScore.toDouble()
	}

	private fun directions(current: Position, memory: Matrix<Double>): Sequence<Direction>
	{
		return Direction.values().asSequence()
				.map { it to (current going it) }
				.sortedByDescending { memory[it.second] }
				.map { it.first }
	}

	private fun reachedComputationCapacity(time: TimeTracker, pathSize: Int): Boolean
	{
		if(time.remaining() < 15)
			return true

		if(pathSize >= maxDepth)
			return true

		return false
	}
}

fun maxStackDepth(snakeSize: Int): Int = 16 + (snakeSize * 1.5).toInt()